### ISSUE TITLE
Quest Redesign

### ISSUE THREAD

**Issue by theeneon** [studie-tech/TheNinjaRPG#132]
- [x] Example : https://kirilllive.github.io/tuesday-js/ . Conclusion: Use as source of inspiration, but do not use code directly, as we want a tight linking to current question functionality + want the scenes to be backed by the database and hide as much information as possible until the user has actually taken actions

- [x] Consecutive Quest/Quest Linking: This is basically a feature where a player has to complete a quest before they are able to unlock a next quest. In the quest creation there will be a field call prerequisite and we can select the quest that needs to be completed before they can take this quest. 

- [x] Repeatable Quest, An option to set how many times a player can retake a quest after failing.
- [x] An option for the quest to show up again after the player completes it after a set amount of time(bi-weekly, monthly, weekly)
- [x] An option to set how many times a player can complete a quest again after completing it. 

- [x] Quest Release Date and Ending. The timer doesn't work, would like a timer where we can set when this quest shows up publically and when it goes away. Show the item's timer so people are aware when it goes away. 

- [x] Expring Items: The ability to set when an item will leave the store, two weeks, one month, 3 months.  When the timer hits the item is automatically hidden

- [x] Dialogs are created elsewhere, On the quest objective section we select Dialogs as an objective and a drop-down to select the dialog folder and then the scene that is created with the ability to set sector and coordinates to trigger this Dialog. 

- [x] #533
- [x] Ability to Set multiple AIs to spawn into a combat instead of just one. 

# Assets
- [x] Ability to create background scenes in manual/asset
- [x] Ability to create characters in manual/asset

# In The Quest creator
The quest flow will be based off "objective" nodes, e.g.

![Image](https://github.com/user-attachments/assets/2407fd1e-c6e6-4756-9c29-2ed5f84b9784)

To realise the wanted system, we need to introduce some changes to current objectives as well as introduce new objectives:

Updates to all objectives
- [x] Background Image
- [x] One or more characters
- [x] Add fail description 
- [x] Music - I've added overall music to the game, but I don't like the music of separate music per quest, and I don't know how to handle it with e.g. multiple quests.
- [x] Item Drops:

New Dialog Objective:
- [x] Text for dialog
- [x] Multiple answers, directing to one of many new objective. This is where the quest flow can split.

Required new "instant" objectives:

- [x] Fail Quest
- [x] Collect Item (Dialog + reward)
- [x] Jutsu (Dialog + reward)
- [x] Bloodline 
- [x] Complete Quest
- [x] Start Battle
- [x] Start new quest

The background will take over the scene, this can probably be done in a pop up similar to the level up pop up that we get.
Characters will then show left or right or in some  cases both left and right or all 3 in the center. I think we have have upload sections for each or decide where each character is placed after uploading the image?
Css for the animations? Writing text effect?

**### Quest Bugs:**

- [x] When the start battle Objective is completed it doesn't progress to the objective after it, have to use the defeat opponent after the start battle for the objective to proceed
- [x] When multiple AIs are added they see each other as enemies instead of allies
- [x] Summons do not work with "failed to create summon" error
- [x] Dialog options are overlaping with descriptions:

![Image](https://github.com/user-attachments/assets/9c33cd68-3f20-4277-a823-258cb8919154)

**### New Quest Objectives:**

- [x] Delivery Objective: Delivery Item to AI at Location and can receive an item from that AI using the new Quest Interface
- [x] Defeat a certain number of AI based on random encounters
- [x] Puzzle Pieces: Collect these pieces of items together to form a full item. Terr Thoughts: This should not be part of the quest system, but rather be based on the collect_item objective, and then I'm guessing the crafting system will enable the "merging" of 2+ items into another, enabling us to create a functionality like this.
- [x] If a Dialog objective comes next have it automatically show up without the player having to move again. 
- [x] Ability to mark an Objective as Optional  so the quest can be completed even if it's not done. RESPONSE: What happens if the second dialog is optional; you go straight to reset, or to win? What if the start_battle is optional; you go straight to dialog, but then there's no battle. What if any end-objective is optional; you win instantly, but never get to try. The way I see it there is a LOT of edge cases we'd have to figure out how to deal with, and simply using dialog options is a simpler way to give "optional" sub-quests; i.e. give the user the option to go straight for a small battle, or go on a side-quest/branch picking up something, and then linking to the final battle. So yea, I think dialog objectives are the way to do optional thing
- [x] Add Reset on Death option if the player dies to the Quest AI they have to restart  the quest from the first objective
- [x] Rename objective. NOTE: This is not trivial, for multiple reasons, as this is the main link between quests. I've reduced the ID sizes for any future objective to be 5 characters instead of the long ones, which hopefully helps. Also added tooltips on the nodes

**Comment by rauf543** [comment]
Proposal for Quest System Enhancements

I would like to work on the following enhancements to the quest system over the next few weeks. Before proceeding, I want to confirm if this direction aligns with the team’s vision and ensure no one is already working on it. Below is a detailed breakdown of the proposed changes:

1. Dialogue Functionality

Objective: Introduce a dialogue system to improve interaction depth and allow for dynamic quest progression.
	•	Implementation Plan:
	•	Add a new inDialogue state to track dialogue status.
	•	Modify the user data schema to include a dialogueID.
	•	Create an activeDialogues schema to manage all active dialogues.
	•	Dialogues will mimic battle logic:
	•	Users are sent into a dialogue session identified by dialogueID.
	•	Dialogues will have multiple outcomes (e.g., answering Yes/No questions).
	•	Future Flexibility:
	•	Introduce variable quest rewards based on dialogue choices (e.g., different rewards depending on the answers to specific questions).

2. Consecutive Quests and Quest Linking

Objective: Enable multi-stage quest chains where the completion of one quest unlocks the next.
	•	Implementation Plan:
	•	Allow quests to be assigned as rewards for completing other quests.
	•	Example:
	•	Quest 1 completion triggers the addition of Quest 2 to the user’s account.
	•	Prevent users from obtaining Quest 2 through other means (e.g., locations like Wake Island).
	•	Future Flexibility:
	•	Dynamic linking based on dialogue choices (e.g., specific quest branches depending on user input).

3. Repeatable Quests and Time-Limited Quests

Objective: Enhance quest variety with repeatable and time-limited quests.
	•	Implementation Plan:
	•	Introduce support for:
	•	Repeatable quests.
	•	Quests with release dates and expiration dates.
	•	Priority: These changes are categorized as quality-of-life improvements and will be implemented after core functional updates.

4. Quest Inventory Management

Objective: Create a robust system for managing items associated with quests.
	•	Implementation Plan:
	•	Investigate the current inventory and item handling systems.
	•	Propose a quest inventory linked to specific quests.
	•	Automatically delete inventory items when the associated quest is completed or deleted.
	•	Further refinement will depend on a deeper analysis of the existing codebase.

5. Quest Content Management System (CMS)

Objective: Adapt the existing quest CMS to support the new features.
	•	Implementation Plan:
	•	Modify the CMS to accommodate dialogue logic and linked quest chains.
	•	This update will likely be developed in parallel with the dialogue system to ensure seamless integration.

Next Steps

	1.	Confirmation: Please confirm if:
	•	No one else is already working on these features.
	•	The proposed direction aligns with the broader roadmap.
	2.	Feedback: Let me know if there are adjustments or other considerations I should incorporate.

This plan breaks down the implementation into clear stages, balancing functionality improvements with long-term flexibility. I look forward to your feedback!

**Comment by theeneon** [comment]
50 |   // User Jutsus & items
  51 |   const { data: userJutsus, isFetching: l1 } = api.jutsu.getUserJutsus.useQuery(
> 52 |     getFilter(state),
     |     ^
  53 |     { enabled: !!userData },
  54 |   );
  55 |   const { data: userItems, isFetching: l2 } = api.item.getUserItems.useQuery(
Static worker exited with code: 1 and signal: null
Error: Process completed with exit code 1.

@openhands-agent

**Comment by TheNinja-RPG** [comment]
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the issue! You can monitor the progress [here](https://github.com/MathiasGruber/TheNinjaRPG/actions/runs/13045773978).

**Comment by TheNinja-RPG** [comment]
A potential fix has been generated and a draft PR #356 has been created. Please review the changes.

**Comment by Phrosfire** [comment]
@MathiasGruber I did do a little looking into the scene building that the event team would like.  I did some testing, and found that we can do a vercel deployment of a simple github repo containing just quest scenes.  The ContentSecurityPolicy has to be updated to allow embedding from the base vercel url we would be using for it.  I did a test where I added vercel to the CSP and the embed worked flawlessly.

If this is something you are open to doing for the time being, I can either build the repo or you can.  It would be best to make it a private repo and only share to those who should have it.  I can go over with the event team how to add their scenes to the repo and navigate to them through the vercel deployment so they can do the embed.

The example provided by Neon (https://kirilllive.github.io/tuesday-js/) provides all the necessary files already contained in a zip folder.  It would just be a matter of unzipping the folder and dropping it into the github repo.  We could realistically put it in the regular TNR repo, but it would require a new PR everytime they want to add something interactive to a quest.

**Comment by MathiasGruber** [comment]
I'd like to do this, and will likely start it as soon as we've cleared some more of the bugs and immidiate needs for the users :)

**Comment by theeneon** [comment]
Am not sure where to put this but somethings that am seeing that i think you should probably do before you go any further on some stuff:

Show the main background and characters on the logbook screen

![Image](https://github.com/user-attachments/assets/3721c35b-0bc3-44d9-ab3d-c5a1bd67eef5)

Don't let objective backgrounds overwite the main quest background so in this screenshot I just accepted the quest, the background that should be showing up should be the one that is linked in the main quest body not the objective:

![Image](https://github.com/user-attachments/assets/65b9c372-908f-4d7b-83cb-fee361606849)

The background and characters for that objective should show up once I get to the objective. 

I also notice that you're placing everything within the Logbook, I think it would be better to give each objective a pop up modal, it would better. 

So if I go to the objective location, a Pop up happens, similar to the image above and you're able to see the background image, the diaglog the options etc. 

Right now when you go to an objective location on the sectors you just do the whatever it is and it shows up as a notification but it would be better if that was just a modal pop up instead

Basically the workflow would like this:

Go to Location>Modal Pop up read through the scene and dialog>take objective(so fight if you're fighting, collect item if you're collecting item , etc)  When you fail or success the modal pops up with the fail or success dialog with the background image there>move on to the next objective.

**Comment by theeneon** [comment]
For the assets page, we need Folders. 

Create Folder>Upload Asset

When on the quest objective, we select Folder First>then select asset.

This should help us Organaize everything and we can probably apply this to AIs and Items as well.

**Comment by MathiasGruber** [comment]
All done, closing this one up 🥳
